codegen/src/main.rs

```diff
@@ -26,41 +28,98 @@ fn main() {
 
-    let in_file: PathBuf = matches.value_of("INPUT")
-        .map(|a| a.to_string().into()).unwrap();
+    let in_files = path_vec(values_t!(matches,"INPUT",String));
+    if in_files.is_empty() {
+        bail!("no .proto files provided!");
+    }
+
+    for f in &in_files {
+        if ! f.exists() {
+            bail!(format!("input file {:?} does not exist", f));
+        }
+    }
+
+    let mut include_path = path_vec(values_t!(matches,"INCLUDE_PATH",String));
+    let default = PathBuf::from(".");
+    if include_path.is_empty() || ! include_path.contains(&default) {
+        include_path.push(default);
+    }
+
+    if in_files.len() > 1 && matches.value_of("OUTPUT").is_some() {
+        bail!("--output only allowed for single input file");
+    }
+
+    for in_file in in_files {
+        let mut out_file = in_file.with_extension("rs");
+
+         if let Some(ofile) = matches.value_of("OUTPUT") {
+            out_file = PathBuf::from(ofile);
+        }
 
-    let out_file: PathBuf = matches.value_of("OUTPUT")
-        .map(|a| a.to_string().into())
-        .unwrap_or(in_file.with_extension("rs"));
+        if let Some(dir) = matches.value_of("OUTPUT_DIR") {
+            let mut directory = PathBuf::from(dir);
+            if ! directory.is_dir() { // we can create? But only last dir
+                bail!(format!("output directory {:?} does not exist", directory));
+            }
+            directory.push(out_file.file_name().unwrap());
+            out_file = directory;
+        }
 
-    let config = Config {
-        in_file: in_file,
-        out_file: out_file,
-        single_module: matches.is_present("SINGLE_MOD")
-    };
+        let config = Config {
+            in_file: in_file,
+            out_file: out_file,
+            single_module: matches.is_present("SINGLE_MOD"),
+            import_search_path: include_path.clone(),
+            no_output: matches.is_present("NO_OUTPUT"),
+        };
 
-    FileDescriptor::write_proto(&config)
-        .expect(&format!("Could not convert {} into {}",
-                         config.in_file.display(), config.out_file.display()));
+        FileDescriptor::write_proto(&config)
+            .chain_err(|| format!("Could not convert {} into {}",
+                             config.in_file.display(), config.out_file.display()))?;
 
+    }
+    Ok(())
 }
 
-fn extension_matches<P: AsRef<Path>>(path: P, expected: &str) -> Result<(), String> {
+fn extension_matches<P: AsRef<Path>>(path: P, expected: &str) -> std::result::Result<(), String> {
     match path.as_ref().extension() {
         Some(x) if x == expected => Ok(()),
         Some(x) => Err(format!("Expected path with extension '{}', not: '{}'", expected, x.to_string_lossy())),
         None => Err(format!("Expected path with extension '{}'", expected)),
     }
 }
+
+fn path_vec(maybe_vec: std::result::Result<Vec<String>,clap::Error>) -> Vec<PathBuf> {
+    maybe_vec.unwrap_or(Vec::new()).iter().map(|s| s.into()).collect()
```

Comment:
Sure. I thought it would be premature optimization to worry about a few vectors at this point ;)